// @nrwl/workspace/src/command-line/affected-apps.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const affected_apps_1 = require("./affected-apps");
const deps_calculator_1 = require("./deps-calculator");
const fs = require("fs");
describe('affected-apps', () => {
    let deps;
    let projects;
    beforeEach(() => {
        spyOn(fs, 'writeFileSync');
        deps = {
            app1Name: [],
            app2Name: [],
            lib1Name: [],
            lib2Name: []
        };
        projects = [
            {
                name: 'app1Name',
                root: 'apps/app1',
                files: ['apps/app1/app1.ts'],
                fileMTimes: {
                    'apps/app1/app1.ts': 1
                },
                tags: [],
                implicitDependencies: [],
                architect: {},
                type: affected_apps_1.ProjectType.app
            },
            {
                name: 'app2Name',
                root: 'apps/app2',
                files: ['apps/app2/app2.ts'],
                fileMTimes: {
                    'apps/app2/app2.ts': 1
                },
                tags: [],
                implicitDependencies: [],
                architect: {},
                type: affected_apps_1.ProjectType.app
            },
            {
                name: 'lib1Name',
                root: 'libs/lib1',
                files: ['libs/lib1/lib1.ts'],
                fileMTimes: {
                    'libs/lib1/lib1.ts': 1
                },
                tags: [],
                implicitDependencies: [],
                architect: {},
                type: affected_apps_1.ProjectType.lib
            },
            {
                name: 'lib2Name',
                root: 'libs/lib2',
                files: ['libs/lib2/lib2.ts'],
                fileMTimes: {
                    'libs/lib2/lib2.ts': 1
                },
                tags: [],
                implicitDependencies: [],
                architect: {},
                type: affected_apps_1.ProjectType.lib
            }
        ];
    });
    describe('affectedAppNames', () => {
        it('should return the list of affected apps', () => {
            deps = Object.assign({}, deps, { app1Name: [
                    {
                        projectName: 'lib1Name',
                        type: deps_calculator_1.DependencyType.es6Import
                    }
                ] });
            const affected = affected_apps_1.affectedAppNames(projects, deps, ['lib1Name']);
            expect(affected).toEqual(['app1Name']);
        });
        it('should handle circular dependencies', () => {
            deps = Object.assign({}, deps, { app1Name: [
                    {
                        projectName: 'app2Name',
                        type: deps_calculator_1.DependencyType.es6Import
                    }
                ], app2Name: [
                    {
                        projectName: 'app1Name',
                        type: deps_calculator_1.DependencyType.es6Import
                    }
                ] });
            const affected = affected_apps_1.affectedAppNames(projects, deps, ['app1Name']);
            expect(affected).toEqual(['app1Name', 'app2Name']);
        });
    });
});
